ipmanager/hetznerFloatingIPConfigurer.go

```diff
@@ -0,0 +1,296 @@
+	*IPConfiguration
+	cachedState  int
+	serverID     int64
+	lastAPICheck time.Time
+	verbose      bool
+}
+
+func newHetznerFloatingIPConfigurer(config *IPConfiguration, verbose bool) (*HetznerFloatingIPConfigurer, error) {
+	c := &HetznerFloatingIPConfigurer{
+		IPConfiguration: config,
+		cachedState:     unknown,
+		serverID:        0,
+		lastAPICheck:    time.Unix(0, 0),
+		verbose:         verbose}
+
+	return c, nil
+}
+
+/**
+ * In order to tell the Hetzner API to route the floating-ip to
+ * this machine, we must attach our own server ID to the API request.
+ */
+
+func (c *HetznerFloatingIPConfigurer) curlQueryFloatingIP(post bool) (string, error) {
+	/**
+	 * The credentials for the API are loaded from a file stored in /etc/hetzner .
+	 */
+	//TODO: make credentialsFile dynamically changeable?
+	credentialsFile := "/etc/hetzner"
+	f, err := os.Open(credentialsFile)
+	if err != nil {
+		log.Println("can't open passwordfile", err)
+		return "", err
+	}
+	defer f.Close()
+
+	/**
+	 * The retrieval of username and password from the file is rather static,
+	 * so the credentials file must conform to the offsets down below perfectly.
+	 */
+	var token string
+	var ip_id string
+	var server_id string
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		line := scanner.Text()
+		switch line[:4] {
+		case "tokn":
+			token     = line[6 : len(line)-1]
+		case "serv":
+			server_id = line[6 : len(line)-1]
+		case "ipid":
+			ip_id     = line[6 : len(line)-1]
+		}
+	}
+	if token == "" || ip_id == "" || server_id == "" {
+		log.Println("Couldn't retrieve API token, IP ID or server ID from file", credentialsFile)
+		return "", errors.New("Couldn't retrieve API token, IP ID or server ID from file")
+	}
+
+	c.serverID, err = strconv.ParseInt(server_id, 10, 64)
+	if err != nil {
+		log.Println("Couldn't convert server id (serv) to number from file", credentialsFile)
+		return "", errors.New("Couldn't convert server id to number from file")
+	}
+
+	/**
+	 * The hetznerFloatingIPConfigurer was copy/paste/modify adopted from the
+	 * hetznerConfigurer. hetznerConfigurer claims that the Hetzner API only
+	 * allows IPv4 connections, and therefore curl is being used instead of
+	 * instead of GO's own http package. I did not verify this for the
+	 * Hetzner Cloud/FloatingIP API so we're also using curl here.
+	 *
+	 * If post is set to true, a failover will be triggered.
+	 * If it is set to false, the current state (i.e. route)
+	 * for the floating-ip will be retrieved.
+	 */
+	var cmd *exec.Cmd
+	if post {
+		log.Printf("my_own_id: %s\n", c.serverID)
+
+		cmd = exec.Command("curl",
+		                   "--ipv4",
+		                   "-X", "POST",
+		                   "-H", "Content-Type: application/json",
+		                   "-H", "Authorization: Bearer "+token,
+		                   "-d", "{\"server\": "+server_id+"}",
+		                   "https://api.hetzner.cloud/v1/floating_ips/"+ip_id+"/actions/assign")
```

Comment:
Maybe with some newfound knowledge about golang, we could start an effort to use golang libraries to send http requests? "How hard can it be?"

The issue in the past was that the Hetzner Robot API limited access to certain IPv4 addresses, this was solved by using curl in ipv4 only mode...

Here's a little draft on how we can use golangs http libary to send requests using only "tcp4" network:
```
package main

import (
	"context"
	"fmt"
	"net"
	"net/http"
	"time"
)

func tcp4DialContext(ctx context.Context, network, address string) (net.Conn, error) {
	DialContext := (&net.Dialer{
		Timeout:   30 * time.Second,
		KeepAlive: 30 * time.Second,
		DualStack: true,
	}).DialContext

	return DialContext(ctx, "tcp4", address)
}

func main() {
	tr := &http.Transport{
		DialContext: tcp4DialContext}
	client := &http.Client{Transport: tr}
	resp, err := client.Get("http://example.com")

	if err != nil {
		fmt.Errorf("something wrong: %e", err)
	} else {
		fmt.Printf("Here's the response:\n%v\n", resp)
	}
}
```

But before we go ahead and try using golang http for this, could you please check that the restriction for certain IPv4 addresses is still there, both with hetzner robot api and hetzner cloud api? Otherwise, we could drop all the IPv4-only shenanigans completely...

Reply:
> But before we go ahead and try using golang http for this, could you please check that the restriction for certain IPv4 addresses is still there, both with hetzner robot api and hetzner cloud api?

You still *must* whitelist the IPs that are allowed to access the Hetzner Robot API and trigger IP routing failovers in the Hetzner web Robot Interface. But you *can* whitelist IPv6 addresses (I have tested this. However I did *not* test if actually triggering the routing failover from an IPv6 IP address works).

Reply:
Also, FYI: https://docs.hetzner.com/cloud/api/faq : "Do you sponsor efforts to integrate Hetzner Cloud into Open Source libs?" "Yes we do: If you are developing integrations based on our API and your product is Open Source you may be eligible for a free one time €50 (excl. VAT) credit on your account"

Reply:
So for the Robot API this restriction is (theoretically) linked to the way the "owner" of the servers sets up the whitelists? Hmm...

Maybe we should create a branch where the curl stuff is replaced, so we can experiment whether it can work with IPv6 as well?

Good to know that such a think exists for the cloud API.
But I guess to do proper testing with the Robot API, one would need to have one or two servers, and since those have high setup costs, I don't think it's feasible for us to continually provide tests against the Hetzner Robot API....

Reply:
> But I guess to do proper testing with the Robot API, one would need to have one or two servers, and since those have high setup costs, I don't think it's feasible for us to continually provide tests against the Hetzner Robot API....

Together with Hetzner's sponsoring I am sure this could eventually be setup as you suggest: 2 servers with cybertec and us (f.ex.) having the credentials to the servers. But we're way ahead of reality here I think: as of now there is zero testing infrastructure in the git repo. So that would need to come first I think?

An alternative approach would be to have a set of users that you can ping for tests of new releases. We do run vip-manager in production so we sure can test the Robot API approach (we won't be productively using the Cluster variant), however I won't be able to give any time guarantees as how quickly we'll be able to test it. Given enough people willing to test, this would not be a problem?